### PR TITLE
Basic CSP parsing and representation.

### PR DESCRIPTION
**Pull request by morlovich**
This is basically the first step, getting the AST and big pieces parsed. The next step would be implementing 
the various "is this URL permitted?" and "is inlining permitted" algorithms from the spec.

### CODE REVIEW

**Comment by morlovich** [comment]
(Had a typo in a unit test which I missed because I had a typo in the test *name* as well, so my test filter excluded it).

**Review by oschaaf — APPROVED** [review]
Aside from the typo this looks good to me.

File: net/instaweb/rewriter/csp.cc [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+    input->remove_prefix(1);
+  }
+
+  while (input->ends_with(" ") || input->ends_with("\t")) {
+    input->remove_suffix(1);
+  }
+}
+
+char Last(StringPiece input) {
+  DCHECK(!input.empty());
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp.cc:43]: should this be a CHECK?   Surely nothing good will come of this if input.empty() is true.

File: net/instaweb/rewriter/csp_directive.gperf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+"connect-src", CspDirective::kConnectSrc
+"default-src", CspDirective::kDefaultSrc
+"frame-src", CspDirective::kFrameSrc
+"img-src", CspDirective::kImgSrc
+"script-src", CspDirective::kScriptSrc
+"style-src", CspDirective::kStyleSrc
+"base-uri", CspDirective::kBaseUri
+%%
+
+CspDirective LookupCspDirective(const StringPiece& name) {
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp_directive.gperf]: StringPiece

File: net/instaweb/rewriter/csp_directive.gperf [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include "net/instaweb/rewriter/public/csp_directive.h"
+#include "pagespeed/kernel/base/string_util.h"
+
+namespace net_instaweb {
+%}
+%compare-strncmp
+%define class-name CspDirectiveMapper
+%define lookup-function-name Lookup
+%define word-array-name CspDirectiveNameTable
+%global-table
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp_directive.gperf:15]: It is not obvious why this deserves the programming overhead of a gperf table, as there is more boilerplate here than actual table.  But I don't really have a problem with it.

File: net/instaweb/rewriter/csp_test.cc [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+TEST(CspParseTest, Empty) {
+  std::unique_ptr<CspPolicy> policy(CspPolicy::Parse(
+      "   "));
+  EXPECT_EQ(policy, nullptr);
+}
+
+TEST(CspParseTest, Basic) {
+  std::unique_ptr<CspPolicy> policy(CspPolicy::Parse(
+    "default-src *; script-src 'unsafe-inline' 'unsafe-eval'"));
+  ASSERT_NE(policy, nullptr);
```
Comment (jmarantz) [inline net/instaweb/rewriter/csp_test.cc]: I find error messages quicker to grok when you spell this:
   ASSERT_TRUE(policy != nullptr);
but this is fine too.

**Review by jmarantz — APPROVED** [review]
Looks nice; just some style comments.

File: net/instaweb/rewriter/public/csp_directive.h [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include <vector>
+
+#include "pagespeed/kernel/base/string.h"
+#include "pagespeed/kernel/base/string_util.h"
+
+namespace net_instaweb {
+
+struct CspDirectiveInfo;
+
+// Directives mentionned in the spec that we care (and comments for those
```
[the remark this answers is not in the document]
Comment (morlovich) [reply net/instaweb/rewriter/public/csp_directive.h]: Done

File: net/instaweb/rewriter/csp_test.cc (hunk 4 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp_test.cc]: Done.

File: net/instaweb/rewriter/csp_directive.gperf (hunk 2 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp_directive.gperf]: Done

File: net/instaweb/rewriter/csp_directive.gperf (hunk 3 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp_directive.gperf:15]: Possibly not, I just wanted flexibility before I sorted out exactly what portion of the spec we care about. 
(Actually there are other specs that may be relevant, this thing is modular...)

File: net/instaweb/rewriter/csp.cc (hunk 1 above)
Comment (morlovich) [reply net/instaweb/rewriter/csp.cc:43]: That just makes it look painfully slow, which is probably just premature optimization.

Maybe I'll just fuzz this thing. It shouldn't be too hard to get it to build in a chrome checkout..

File: net/instaweb/rewriter/csp.cc [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+#include "net/instaweb/rewriter/public/csp_directive.h"
+
+namespace net_instaweb {
+
+namespace {
+
+void TrimCspWhitespace(StringPiece* input) {
+  // AKA RWS in HTTP spec, which of course isn't the HTML notion of whitespace
+  // that TrimWhitespace uses.
+  while (!input->empty() && ((*input)[0] == ' ' || (*input)[1] == '\t')) {
```
Comment (morlovich) [inline net/instaweb/rewriter/csp.cc]: Err, should probably fix + test that [1] too....

**Pull request merged**